[TECH] Ajouter une API de suppression de compte utilisateur (PIX-14910)

er-lim opened the pull request:
## :fallen_leaf: Problème

Il est nécessaire d'implémenter la partie back/API de la fonctionnalité de suppression de compte en autonomie.

<img width="1417" alt="Capture d’écran 2024-11-21 à 11 41 29" src="https://github.com/user-attachments/assets/dcbf27ea-ef1e-40e4-8257-997b06b47211">

## :chestnut: Proposition

### Une nouvelle route

On a choisi une nouvelle route `/api/users/me`, qui ne contient aucun paramètre `userId`, pour ne permettre à l'utilisateur authentifié de ne supprimer que son propre compte. Ce choix augmente la sécurité tout en réduisant la complexité du code au minimum.

### Cas d'erreur et de réussite

* Si la suppression de compte n'est pas possible alors la route envoie une erreur `403`
* Si la suppression de compte est possible alors le usecase d’anonymisation est exécuté

### Choix de nommage

* Pour le bounded context _Privacy_ on a appelé les usecases et fonctions suivant le nom *anonymizeUser* car c'est techniquement ce qui est réalisé.
* Pour le bounded context _Identity Access Management_ on a appelé les usecases et fonctions suivant le nom *selfDeleteUserAccount* car on est dans une fonctionnalité métier qui du point de vue de l'utilisateur est de supprimer son compte. L'utilisateur ne sait pas que techniquement toutes ses données vont en fait être anonymisées, et d'ailleurs dans le futur on ne sait pas si la fonctionnalité métier _selfDeleteUserAccount_ ne supprimerait pas purement et simplement certaines données

### Stratégie de tests

On n'a fait aucun test pour le controller, car conformément à la stratégie de tests : 
* les tests (d'intégration et d'acceptance) sur la route testent déjà tout
* le code du controller est minimaliste

### Création d'une nouvelle API canSelfDeleteAccount et factorisation

On a créé une nouvelle API `canSelfDeleteAccount` et on a factorisé en utilisant cette nouvelle API pour le usecase existant `api/src/identity-access-management/domain/usecases/anonymize-user.usecase.js`.

## :jack_o_lantern: Remarques

RAS

## :wood: Pour tester

Pour tester en local il y a des actions à réaliser au préalable :

1. Faire un build de l'Audit Logger pour que le code TypeScript génère du JavaScript à jour :

```shell
cd api/audit-logger/
npm run build
```

2. Démarrer l'Audit Logger

```shell
cd api/audit-logger/
npm start
```

3. Lancer les jobs

```shell
cd api/
npm run start:job
```

---

1. Créer un nouveau compte utilisateur (et qui donc n'aura aucune participation à Pix et pourra être supprimé sans contrainte)
2. Se déconnecter
3. Se connecter en super admin sur Pix Admin
4. Aller sur la page _Utilisateurs_
5. Rechercher l’utilisateur créé dans l'étape _1._
6. Actionner le bouton `Anonymiser cet utilisateur`
7. Vérifier que l'utilisateur a bien été anonymisé
8. Vérifier que l'événement de l'anonymisation est bien présent dans l'AuditLogger

er-lim committed:
feat(api): add self delete user account usecase

in identity-access-management bounded context

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr10606.review.pix.fr)
  * [App (.org)](https://app-pr10606.review.pix.org)
  * [Orga (.fr)](https://orga-pr10606.review.pix.fr)
  * [Orga (.org)](https://orga-pr10606.review.pix.org)
  * [Certif (.fr)](https://certif-pr10606.review.pix.fr)
  * [Certif (.org)](https://certif-pr10606.review.pix.org)
  * [Junior](https://junior-pr10606.review.pix.fr)
  * [Admin](https://admin-pr10606.review.pix.fr)
  * [API](https://api-pr10606.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr10606.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr10606/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr10606/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr10606/environment)

lego-technix committed:
feat(api): add anonymize-user.usecase to Privacy bounded context

lego-technix committed:
feat(api): add anonymizeUser function to Privacy bounded context users-api.js

lego-technix committed:
refactor(api): remove now unused case when there is no updatedByUserId

There was no updatedByUserId when the backfill-anonymized-users.js script was
used. This script has now been removed.

lego-technix committed:
refactor(api): remove preventAuditLogging now unused property

lego-technix committed:
feat(api): new allowed values for UserAnonymizedEventLoggingJob

lego-technix committed:
test(api): fix a missing DDD context in the test describe

lego-technix committed:
test(api): transform users-api unit tests into integration tests to follow standards

lego-technix committed:
feat(api): add self delete user account route and controller

in identity-access-management bounded context

lego-technix committed:
refactor(api): leverage new privacyUsersApiRepository.anonymizeUser API

er-lim commented on api/tests/privacy/integration/application/api/users-api.test.js:
```diff
@@ -0,0 +1,23 @@
+      const userId = databaseBuilder.factory.buildUser().id;
+      await databaseBuilder.commit();
+
+      const usecaseStub = sinon.stub(usecases, 'canSelfDeleteAccount');
```
**Question** Je me pose des questions sur l'utilité d'un test d'intégration ici . Les données insérées ne sont pas utilisées ce qui fait que je ne vois pas de différence avec le test unitaire 😄 

Est-ce que c'est pour préparer la suite que ça a été fait comme ça ?

lego-technix replied:
Ma stratégie a uniquement été de faire des tests d'intégration chaque fois que c'était possible (du moment que ce n'était pas des modèles) car c'est ce qui se rapproche le plus possible de la réalité sans être aussi lourd que de l'acceptance à réserver pour un cas nominal/les cas nominaux. C'est ce que j'ai compris de la stratégie de tests :confused: 

Je suggère de rediscuter tous les points concernant les tests lors d'une discussion globale sur la stratégie de tests.

bpetetot commented on api/tests/identity-access-management/acceptance/application/user/user.route.test.js:
```diff
@@ -895,6 +896,38 @@ describe('Acceptance | Identity Access Management | Application | Route | User',
+      const options = {
+        method: 'DELETE',
+        url: '/api/users/me',
+        payload: {},
```
Le payload n'est pas nécessaire ici.
```suggestion
```

bpetetot commented on api/tests/identity-access-management/integration/domain/usecases/self-delete-user-account.usecase.test.js:
```diff
@@ -0,0 +1,38 @@
+
+      sinon.stub(config.featureToggles, 'isSelfAccountDeletionEnabled').value(true);
+
+      // when
+      const result = await usecases.selfDeleteUserAccount({ userId });
+
+      // then
+      expect(result).to.be.undefined;
```
Pour mieux coller à l'intitulé du test, je propose ce type d'assertion:
```js
// when / then
await expect(usecases.selfDeleteUserAccount({userId})).to.not.be.rejectedWith(ForbiddenAccess);
```

À noter qu'il utilise [`chai-as-promised`](https://github.com/chaijs/chai-as-promised) une librairie `chai` récemment ajouter dans le projet `chai` qui permet de réaliser des assertions avancées sur les promesses.

bpetetot commented on api/tests/identity-access-management/integration/domain/usecases/self-delete-user-account.usecase.test.js:
```diff
@@ -0,0 +1,38 @@
+
+      sinon.stub(config.featureToggles, 'isSelfAccountDeletionEnabled').value(false);
+
+      // when
+      const error = await catchErr(usecases.selfDeleteUserAccount)({ userId });
+
+      // then
+      expect(error).to.be.instanceOf(ForbiddenAccess);
```
Pour info, l'utilisation du helper maison `catchErr` ne devrait plus avoir lieu d'être avec l'intégration de [`chai-as-promised`](https://github.com/chaijs/chai-as-promised) dans `chai`.

```js
// when / then
await expect(usecases.selfDeleteUserAccount({userId})).to.be.rejectedWith(ForbiddenAccess);
```

C'est peut-être une API à présenter plus globalement aux devs Pix.

bpetetot commented on api/tests/identity-access-management/integration/infrastructure/repositories/privacy-users-api.repository.test.js:
À (éventuellement) revoir dans notre stratégie de test, mais pour les repositories qui sont dédiés aux APIs externes d'un contexte, on devrait privilégier soit des tests unitaires, soit des tests d'intégration avec les APIs stubbées. Car du point de vue du contexte, on peut les considérer comme des services tiers dont on ne connait pas le modèle de données (par exemple, c'est comme si on souhaitée une API d'un SSO, de Salesforce, Slack...). On veut s'assurer de l'appeler avec les bons paramètres. À discuter avec l'équipe.

Pour ce test d'intégration, qui est très simple, on peut le laisser comme ça pour le moment.

bpetetot commented on api/tests/privacy/integration/application/api/users-api.test.js:
Pour les tests d'API, on va toujours privilégier des tests unitaires, car ce ne sont que des fonctions JS qui font que des appels à des use-cases à partir des paramètres données en paramètres et retourne (ou non un résultat).

La logique métier, et donc les tests d'intégration, sont réalisés au niveau du use case.

D'ailleurs, ce que tu as écris sont bien des tests unitaires des fonctions de l'API. Tu as stubbé les usecases qui sont les seuls effets de bord de l'API. Donc il suffit juste de déplacer ce fichier dans les tests unitaires et enlever la création des utilisateurs en base (car ils ne sont jamais utilisés dans l'exécution du test)

bpetetot commented on api/src/privacy/application/api/users-api.js:
```diff
@@ -1,12 +1,18 @@
 import { usecases } from '../../domain/usecases/index.js';
 
+const anonymizeUser = async ({ userId, anonymizedByUserId, anonymizedByUserRole, client }) => {
+  return usecases.anonymizeUser({ userId, anonymizedByUserId, anonymizedByUserRole, client });
+};
```
Comme il s'agit d'une API utilisée à l'extérieur du contexte, il faudrait la documenter avec une JS doc

bpetetot commented on api/src/privacy/domain/usecases/anonymize-user.usecase.js:
```diff
@@ -0,0 +1,93 @@
+import { config } from '../../../shared/config.js';
+import { withTransaction } from '../../../shared/domain/DomainTransaction.js';
+
+/**
+ * @param params
+ * @param{string} params.userId
+ * @param{string} params.anonymizedByUserId
+ * @param{string} params.client
+ * @param{UserRepository} params.userRepository
+ * @param{AuthenticationMethodRepository} params.authenticationMethodRepository
+ * @param{MembershipRepository} params.membershipRepository
+ * @param{CertificationCenterMembershipRepository} params.certificationCenterMembershipRepository
+ * @param{OrganizationLearnerRepository} params.organizationLearnerRepository
+ * @param{RefreshTokenRepository} params.refreshTokenRepository
+ * @param{ResetPasswordDemandRepository} params.resetPasswordDemandRepository
+ * @param{UserLoginRepository} params.userLoginRepository
+ * @param{AdminMemberRepository} params.adminMemberRepository
+ * @param{UserAnonymizedEventLoggingJobRepository} params.userAnonymizedEventLoggingJobRepository
+ * @returns {Promise<null>}
+ */
```
Mise à jour de la JS doc
```suggestion
/**
 * @param params
 * @param{string} params.userId
 * @param{string} params.anonymizedByUserId
 * @param{string} params.anonymizedByUserRole
 * @param{string} params.client
 * @param{UserRepository} params.userRepository
 * @param{AuthenticationMethodRepository} params.authenticationMethodRepository
 * @param{MembershipRepository} params.membershipRepository
 * @param{CertificationCenterMembershipRepository} params.certificationCenterMembershipRepository
 * @param{OrganizationLearnerRepository} params.organizationLearnerRepository
 * @param{RefreshTokenRepository} params.refreshTokenRepository
 * @param{ResetPasswordDemandRepository} params.resetPasswordDemandRepository
 * @param{UserLoginRepository} params.userLoginRepository
 * @param{UserAnonymizedEventLoggingJobRepository} params.userAnonymizedEventLoggingJobRepository
 * @returns {Promise<void>}
 */
```

bpetetot commented on api/src/privacy/domain/usecases/anonymize-user.usecase.js:
```diff
@@ -0,0 +1,93 @@
+      }),
+    );
+  }
+  return null;
```
Cette fonction ne devrait rien retourner.
```suggestion
```

lego-technix committed:
test(api): use dependency inject for usecase integration tests + use chai-as-promised

lego-technix committed:
feat(api): add anonymizeUser function to Identity-access-management bounded context privacy-users-api.repository.js

lego-technix commented:
Je viens de finir de mettre à jour les tests suivant la nouvelle mouture de notre stratégie de tests. C'est prêt pour une nouvelle revue.

Suite à cette clarification j'ai également simplifié la description de la PR.

bpetetot reviewed (approved):
👍 Revu et testé

EmmanuelleBonnemay reviewed (approved):
Lu et testé avec succès
